src/core/model/table/dynamic_position_list_index.cpp

```diff
@@ -0,0 +1,200 @@
+                                                   size_t relation_size)
+    : clusters_(std::move(clusters)),
+      size_(size),
+      relation_size_(relation_size),
+      probing_table_cache_() {}
+
+std::unique_ptr<DynamicPositionListIndex> DynamicPositionListIndex::CreateFor(
+        std::vector<std::vector<int>>& data) {
+    ClusterCollection clusters;
+    for (size_t position = 0; position < data.size(); ++position) {
+        std::vector<int> values = data[position];
+        clusters[values].push_back(position);
+    }
+
+    for (auto& [_, cluster] : clusters) {
+        sort(cluster.begin(), cluster.end());
+    }
+
+    return std::make_unique<DynamicPositionListIndex>(std::move(clusters), data.size(),
+                                                      data.size());
+}
+
+void DynamicPositionListIndex::UpdateWith(
+        std::vector<std::pair<std::optional<size_t>, std::vector<int>>>& inserted_records,
+        std::vector<size_t> const& deleted_records_ids) {
+    if (inserted_records.empty() && deleted_records_ids.empty()) {
+        return;
+    }
+    std::map<std::vector<int>, std::vector<int>> index;
+    for (size_t position = 0; position < inserted_records.size(); ++position) {
+        std::vector<int> values = std::move(inserted_records[position].second);
+        if (inserted_records[position].first.has_value()) {
+            index[values].emplace_back(inserted_records[position].first.value());
+        } else {
+            index[values].emplace_back(relation_size_++);
+            ++size_;
+        }
+    }
+    size_ -= deleted_records_ids.size();
+
+    std::vector<bool> is_deleted(deleted_records_ids.size(), false);
+    for (auto it = clusters_.begin(); it != clusters_.end();) {
+        ProcessDeleteToCluster(it->second, deleted_records_ids, is_deleted);
+        if (index.contains(it->first)) {
+            it->second.insert(it->second.end(), index[it->first].begin(), index[it->first].end());
+            index.erase(it->first);
+        }
+
+        if (it->second.empty()) {
+            it = clusters_.erase(it);
+        } else {
+            sort(it->second.begin(), it->second.end());
+            ++it;
+        }
+    }
+
+    for (auto& new_cluster : index) {
+        clusters_.emplace(std::move(new_cluster));
+    }
+}
+
+std::unordered_map<int, unsigned> DynamicPositionListIndex::CreateFrequencies(
+        Cluster const& cluster, std::vector<int> const& probing_table) {
+    std::unordered_map<int, unsigned> frequencies;
+
+    for (int index : cluster) {
+        frequencies[probing_table[index]]++;
+    }
+
+    return frequencies;
+}
+
+std::shared_ptr<std::vector<int> const> DynamicPositionListIndex::CalculateAndGetProbingTable()
+        const {
+    if (probing_table_cache_ != nullptr) return probing_table_cache_;
+    cluster_values_cache_.clear();
+    std::vector<int> probing_table = std::vector<int>(relation_size_);
```

Comment:
`assert(kNullValueId == 0)`

Reply:
where and why? also, we got `static constexpr int kNullValueId = -1;` expression in dynamic_fd_verifier.h

Reply:
Oh, I misunderstood something, then